apollo-federation/src/composition/satisfiability/validation_state.rs

```diff
@@ -64,6 +64,7 @@ struct SubgraphPathContextInfo {
 #[derive(PartialEq, Eq, Hash)]
 pub(super) struct SubgraphContextKey {
     tail_subgraph_name: Arc<str>,
+    tail_type_name: String,
```

Comment:
I just realized we can use `QueryGraphNodeType` directly here.
```suggestion
    tail_type: QueryGraphNodeType,
```
This will be more efficient than `String`.